File: cpp/questions/rob.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int rob(vector<int>& nums) {
        int ns = nums.size();
        std::array<int, 2> dp = {0, 0};
        for (const auto& n: nums) {
            int cr = dp[0] + n;
            dp[0] = std::max(dp[0], dp[1]);
            dp[1] = std::max(cr, dp[1]);
        }
        return std::max(dp[0], dp[1]);
    }
};

File: python/questions/rob.py
from typing import List

class Solution:
    def rob(self, nums: List[int]) -> int:
        ln = len(nums)
        cm = [[0, 0] for _ in range(ln+1)]
        cm[1] = [nums[0], nums[0]]
        for i in range(1, ln):
            cm[i+1][0] = cm[i-1][1] + nums[i]
            cm[i+1][1] = max(cm[i+1][0], cm[i][1])
        return cm[ln][1]
